core/blockchain.go

```diff
@@ -196,6 +196,7 @@ type BlockChain struct {
 
 	quit          chan struct{}  // blockchain quit channel
 	wg            sync.WaitGroup // chain processing wait group for shutting down
+	wgMu          sync.Mutex     // chain processing wait group for shutting down
```

Comment:
Is this mutex intended to make sure the `Wait` happens after the first `Add`? AFAIK `sync.WaitGroup` is thread safe in its use and there shouldn't be any need to syncronize its access, but I may be missing some border case

Reply:
This mutex ensure that wait and add cannot be called concurrently see - https://github.com/golang/go/issues/23842

Reply:
Ah, right, it's thread safe if you call Add when the count is already > 0, but if it's exactly 0 then it's not thread safe. Either a broken API or a weird use by ourselves.

Given that in this particular use case Add should never be called from a different goroutine without the other lock, I think we better encapsulate this in a new type (`DynamicWaitGroup` or whatever name you can think of) and have its own internal mutex, to avoid adding one more reference to the blockchain type, and also making sure that any new code written here cannot accidentally call the Add without the lock.